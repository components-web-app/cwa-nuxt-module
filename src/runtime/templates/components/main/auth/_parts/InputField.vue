<template>
  <div>
    <label
      :for="id"
      class="cwa:block cwa:text-sm cwa:font-medium cwa:leading-6 cwa:text-neutral-300"
    >{{ props.label }}</label>
    <div class="cwa:mt-2">
      <input
        :id="id"
        :name="props.name"
        :type="props.type"
        :autocomplete="props.autocomplete"
        :required="props.required"
        :value="modelValue"
        class="cwa:block cwa:w-full cwa:border-0 cwa:py-1.5 cwa:text-neutral-900 cwa:shadow-sm cwa:ring-1 cwa:ring-inset cwa:ring-neutral-300 cwa:placeholder:text-neutral-400 cwa:focus:ring-2 cwa:focus:ring-inset cwa:focus:ring-neutral-600 cwa:sm:text-sm cwa:sm:leading-6"
        @input="updateModelValue"
      >
      <div v-if="errors">
        <CwaUiAlertWarning
          v-for="(error, index) in errors"
          :key="`error-${id}-${index}`"
          class="cwa:mt-2"
        >
          {{ error }}
        </CwaUiAlertWarning>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { useId } from 'vue'

const emit = defineEmits(['update:modelValue'])

const props = withDefaults(defineProps<{
  name: string
  type?: string
  label: string
  autocomplete?: string
  required?: boolean
  modelValue: string
  errors?: string[]
}>(), {
  type: 'text',
  autocomplete: undefined,
  required: false,
  modelValue: '',
  errors: undefined,
})

const id = useId()

function updateModelValue($event: Event) {
  if (!($event.target instanceof HTMLInputElement)) {
    return
  }
  emit('update:modelValue', $event.target?.value)
}
</script>
