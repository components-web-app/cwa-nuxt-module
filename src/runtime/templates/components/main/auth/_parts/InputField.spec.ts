// @vitest-environment happy-dom
import { describe, expect, test } from 'vitest'
import { shallowMount } from '@vue/test-utils'
import InputField from './InputField.vue'

interface InputMeta {
  name: string
  label: string
  modelValue: string
  type?: string
  autocomplete?: string
  required?: boolean
  errors?: string[]
}

function createWrapper({
  name,
  label,
  modelValue,
  type = 'text',
  autocomplete = 'login',
  required = false,
  errors,
}: InputMeta) {
  return shallowMount(InputField, {
    props: {
      name,
      label,
      modelValue,
      type,
      autocomplete,
      required,
      errors,
    },
    global: {
      renderStubDefaultSlot: true,
    },
  })
}

describe('InputField', () => {
  describe('emits', () => {
    test('should emit updated value', async () => {
      const mockNewValue = 'new input value'
      const wrapper = createWrapper({
        name: 'login',
        label: 'Login',
        modelValue: '',
      })

      const input = wrapper.find('input')

      await input.setValue(mockNewValue)

      expect(wrapper.emitted('update:modelValue')).toEqual([[mockNewValue]])
    })
  })

  describe('props', () => {
    test('native input should accept value from modelValue prop', () => {
      const value = 'mock_user'
      const wrapper = createWrapper({
        name: 'login',
        label: 'Login',
        modelValue: value,
      })

      const input = wrapper.find('input')

      expect(input.element.value).toEqual(value)
    })
  })

  describe('snapshots', () => {
    test('should match snapshot with passed content', () => {
      const wrapper = createWrapper({
        name: 'login',
        label: 'Login',
        modelValue: '',
      })

      expect(wrapper.element).toMatchSnapshot()
    })

    test('should match snapshot IF input is marked as required', () => {
      const wrapper = createWrapper({
        name: 'login',
        label: 'Login',
        modelValue: '',
        required: true,
      })

      expect(wrapper.element).toMatchSnapshot()
    })

    test('should match snapshot IF input is of type password', () => {
      const wrapper = createWrapper({
        name: 'password',
        label: 'Password',
        autocomplete: 'password',
        modelValue: '',
        type: 'password',
      })

      expect(wrapper.element).toMatchSnapshot()
    })

    test('should match snapshot IF errors are present', () => {
      const wrapper = createWrapper({
        name: 'login',
        label: 'Login',
        modelValue: '',
        errors: ['Login should be minimum 6 characters long', 'Login cannot contain whitespaces'],
      })

      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
