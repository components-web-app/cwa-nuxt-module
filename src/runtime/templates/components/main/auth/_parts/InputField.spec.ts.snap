// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputField > snapshots > should match snapshot IF errors are present 1`] = `
<div>
  <label
    class="cwa:block cwa:text-sm cwa:font-medium cwa:leading-6 cwa:text-neutral-300"
    for="v-0"
  >
    Login
  </label>
  <div
    class="cwa:mt-2"
  >
    <input
      autocomplete="login"
      class="cwa:block cwa:w-full cwa:border-0 cwa:py-1.5 cwa:text-neutral-900 cwa:shadow-sm cwa:ring-1 cwa:ring-inset cwa:ring-neutral-300 cwa:placeholder:text-neutral-400 cwa:focus:ring-2 cwa:focus:ring-inset cwa:focus:ring-neutral-600 cwa:sm:text-sm cwa:sm:leading-6"
      id="v-0"
      name="login"
      type="text"
      value=""
    />
    <div>

      <warning-stub
        class="cwa:mt-2"
      >
        Login should be minimum 6 characters long
      </warning-stub>
      <warning-stub
        class="cwa:mt-2"
      >
        Login cannot contain whitespaces
      </warning-stub>

    </div>
  </div>
</div>
`;

exports[`InputField > snapshots > should match snapshot IF input is marked as required 1`] = `
<div>
  <label
    class="cwa:block cwa:text-sm cwa:font-medium cwa:leading-6 cwa:text-neutral-300"
    for="v-0"
  >
    Login
  </label>
  <div
    class="cwa:mt-2"
  >
    <input
      autocomplete="login"
      class="cwa:block cwa:w-full cwa:border-0 cwa:py-1.5 cwa:text-neutral-900 cwa:shadow-sm cwa:ring-1 cwa:ring-inset cwa:ring-neutral-300 cwa:placeholder:text-neutral-400 cwa:focus:ring-2 cwa:focus:ring-inset cwa:focus:ring-neutral-600 cwa:sm:text-sm cwa:sm:leading-6"
      id="v-0"
      name="login"
      required=""
      type="text"
      value=""
    />
    <!--v-if-->
  </div>
</div>
`;

exports[`InputField > snapshots > should match snapshot IF input is of type password 1`] = `
<div>
  <label
    class="cwa:block cwa:text-sm cwa:font-medium cwa:leading-6 cwa:text-neutral-300"
    for="v-0"
  >
    Password
  </label>
  <div
    class="cwa:mt-2"
  >
    <input
      autocomplete="password"
      class="cwa:block cwa:w-full cwa:border-0 cwa:py-1.5 cwa:text-neutral-900 cwa:shadow-sm cwa:ring-1 cwa:ring-inset cwa:ring-neutral-300 cwa:placeholder:text-neutral-400 cwa:focus:ring-2 cwa:focus:ring-inset cwa:focus:ring-neutral-600 cwa:sm:text-sm cwa:sm:leading-6"
      id="v-0"
      name="password"
      type="password"
      value=""
    />
    <!--v-if-->
  </div>
</div>
`;

exports[`InputField > snapshots > should match snapshot with passed content 1`] = `
<div>
  <label
    class="cwa:block cwa:text-sm cwa:font-medium cwa:leading-6 cwa:text-neutral-300"
    for="v-0"
  >
    Login
  </label>
  <div
    class="cwa:mt-2"
  >
    <input
      autocomplete="login"
      class="cwa:block cwa:w-full cwa:border-0 cwa:py-1.5 cwa:text-neutral-900 cwa:shadow-sm cwa:ring-1 cwa:ring-inset cwa:ring-neutral-300 cwa:placeholder:text-neutral-400 cwa:focus:ring-2 cwa:focus:ring-inset cwa:focus:ring-neutral-600 cwa:sm:text-sm cwa:sm:leading-6"
      id="v-0"
      name="login"
      type="text"
      value=""
    />
    <!--v-if-->
  </div>
</div>
`;
